vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,666 @@
+    // filter actions that are not explicitly included
+    // if (!data.include_list.empty() && data.include_list.find((uint32_t)i) == data.include_list.end())
+    //   continue;
+
+    // filter actions chosen by previous slots
+    if (data.exclude_list[i])
+      continue;
+
+    // select the action
+    cb_ex.push_back(data.actions[i]);
+
+    // save the original index from the root multi-example
+    data.origin_index[index++] = (uint32_t)i;
+
+    // remember the index of the chosen action
+    if (is_learn && slot_has_label && i == slot->l.conditional_contextual_bandit.outcome->probabilities[0].action)
+    {
+      // This is used to remove the label later.
+      data.action_with_label = (uint32_t)i;
+      attach_label_to_example(index, data.actions[i], slot->l.conditional_contextual_bandit.outcome, data);
+    }
+  }
+
+  // Must reset this in case the pooled example has stale data here.
+  data.shared->pred.a_s = data.action_score_pool.get_object();
+
+  // Tag can be used for specifying the sampling seed per slot. For it to be used it must be inserted into the shared
+  // example.
+  std::swap(data.shared->tag, slot->tag);
+}
+
+// iterate over slots contained in the multi-example, and for each slot, build a cb example and perform a
+// cb_explore_adf call.
+template <bool is_learn>
+void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
+{
+  clear_all(data);
+  split_multi_example(examples, data);  // split shared, actions and slots
+
+#ifndef NDEBUG
+  sanity_checks<is_learn>(data);
+#endif
+
+  // Stash the CCB labels before rewriting them.
+  data.stored_labels.clear();
+  for (auto ex : examples)
+  {
+    data.stored_labels.push_back({ex->l.conditional_contextual_bandit.type, ex->l.conditional_contextual_bandit.outcome,
+        ex->l.conditional_contextual_bandit.explicit_included_actions});
+  }
+
+  // This will overwrite the labels with CB.
+  create_cb_labels(data);
+
+  // Reset exclusion list for this example.
+  data.exclude_list.assign(data.actions.size(), false);
+
+  auto decision_scores = data.action_scores_pool.get_object();
+
+  // for each slot, re-build the cb example and call cb_explore_adf
+  int slot_id = 0;
+  for (example* slot : data.slots)
+  {
+    // Namespace crossing for slot features.
+    // If the slot example only has the constant namespace, there will be no extra crossing and so skip that logic.
+    if (!(slot->indices.size() == 1 && slot->indices[0] == constant_namespace))
+    {
+      data.generated_interactions.clear();
+      std::copy(data.original_interactions->begin(), data.original_interactions->end(),
+          std::back_inserter(data.generated_interactions));
+      calculate_and_insert_interactions(data.shared, data.actions, data.generated_interactions);
+      data.shared->interactions = &data.generated_interactions;
+      for (auto ex : data.actions)
+      {
+        ex->interactions = &data.generated_interactions;
+      }
+    }
+
+    multi_ex cb_ex;
```

Comment:
Use static thread_local, with clear pattern to avoid allocation here?

For example:
[https://github.com/rajan-chari/vowpal_wabbit/blob/5c21b2dfda90cd8beb1372a0926f4a445dffb708/vowpalwabbit/offset_tree.cc#L128](https://github.com/rajan-chari/vowpal_wabbit/blob/5c21b2dfda90cd8beb1372a0926f4a445dffb708/vowpalwabbit/offset_tree.cc#L128)